### PR TITLE
[MAINT] add script to check duration of test suite

### LINKED ISSUE #4220: speed up test for reporting
I did not realize that by adding some parametrization I may have increased the run time of our test suite.

May need to fixturize some of this so speed it up.

```
2024-01-16T13:43:19.6964134Z 13.22s call     nilearn/reporting/tests/test_html_report.py::test_nifti_maps_masker_report_integer_and_all_displayed_maps[9]
2024-01-16T13:43:19.6964860Z 13.16s call     nilearn/reporting/tests/test_html_report.py::test_nifti_maps_masker_report_integer_and_all_displayed_maps[all]
2024-01-16T13:43:19.6965228Z 13.09s call     nilearn/reporting/tests/test_html_report.py::test_report_empty_fit[NiftiMapsMasker]
2024-01-16T13:43:19.6965684Z 13.09s call     nilearn/reporting/tests/test_html_report.py::test_warning_in_report_after_empty_fit[NiftiMapsMasker]
2024-01-16T13:43:19.6966272Z 13.08s call     nilearn/reporting/tests/test_html_report.py::test_report_empty_fit[MultiNiftiMapsMasker]
2024-01-16T13:43:19.6967074Z 13.06s call     nilearn/reporting/tests/test_html_report.py::test_nifti_maps_masker_report_integer_and_all_displayed_maps[12]
2024-01-16T13:43:19.6968702Z 9.75s call     nilearn/reporting/tests/test_html_report.py::test_multi_nifti_maps_masker_report_warning
2024-01-16T13:43:19.6970190Z 8.87s call     nilearn/reporting/tests/test_html_report.py::test_nifti_maps_masker_report_integer_and_all_displayed_maps[6]
2024-01-16T13:43:19.6970760Z 8.06s call     nilearn/reporting/tests/test_html_report.py::test_nifti_labels_masker_report
2024-01-16T13:43:19.6974957Z 7.06s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[fpr-True]
2024-01-16T13:43:19.6975712Z 7.06s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[bonferroni-True]
2024-01-16T13:43:19.6976391Z 7.02s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[fdr-True]
2024-01-16T13:43:19.6976992Z 6.99s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[fpr-False]
2024-01-16T13:43:19.6977452Z 6.95s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[bonferroni-False]
2024-01-16T13:43:19.6978113Z 6.91s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[fdr-False]
2024-01-16T13:43:19.6978800Z 6.79s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[None-True]
2024-01-16T13:43:19.6979373Z 6.68s call     nilearn/reporting/tests/test_glm_reporter.py::test_slm_reporting[None-False]
2024-01-16T13:43:19.6980005Z 6.53s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[None-True]
2024-01-16T13:43:19.6980418Z 6.52s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[None-False]
2024-01-16T13:43:19.6980964Z 6.50s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[fdr-False]
2024-01-16T13:43:19.6981540Z 6.48s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[bonferroni-False]
2024-01-16T13:43:19.6982158Z 6.46s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[fdr-True]
2024-01-16T13:43:19.6982949Z 6.46s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[fpr-False]
2024-01-16T13:43:19.6983756Z 6.44s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[bonferroni-True]
2024-01-16T13:43:19.6984410Z 6.40s call     nilearn/reporting/tests/test_glm_reporter.py::test_flm_reporting[fpr-True]
```

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to #4220 

Changes proposed in this pull request:

- add script to collect duration of runs of the test suite in github action and plot the results it over time
- mostly for monitoring
- does not go too far back as it as it tracks the workflow that uses tox that is fairly new

### CODE REVIEW

**Comment by Remi-Gau** [comment]
output example

newplot(1)

**Comment by Remi-Gau** [comment]
does not have to go into this repo but handy to have around

**Comment by bthirion** [comment]
> output example
> 
> newplot(1)

Thx. Not that the use of  a unqiue color for different curves makes it barely usable.

**Comment by bthirion** [comment]
OK, you will then create the Github action, I guess ?

**Review by bthirion — APPROVED** [review]
Could not spot any issue.

**Comment by Remi-Gau** [comment]
> OK, you will then create the Github action, I guess ?

yeah to run on command and every month or so.

File: maint_tools/requirements.txt [hunk 1]
```diff
@@ -1,2 +1,6 @@
 docstring_parser
+kaleido
+pandas
+plotly
+request
```
Comment (Remi-Gau) [inline maint_tools/requirements.txt:5]:
```suggestion
requests
```

**Comment by Remi-Gau** [comment]
> Not that the use of a unqiue color for different curves makes it barely usable.

improved this a bit:

- color = python version
- dash = OS
- marker = dependencies

test_runs_timing

**Comment by Remi-Gau** [comment]
> OK, you will then create the Github action, I guess ?

This is done

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Comment by bthirion** [comment]
Happy to hear about other's opinion on this feature.

**Comment by Remi-Gau** [comment]
> Happy to hear about other's opinion on this feature.

As this is not urgent I will wait for a couple of weeks before merging.

**Comment by Remi-Gau** [comment]
Unless I hear anything otherwise I will merge this next week.

File: maint_tools/check_gha_workflow.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+import pandas as pd
+import plotly.express as px
+import requests
+from rich import print
+
+USER = "nilearn"
+REPO = "nilearn"
+
+# Your github username
+USERNAME = "Remi-Gau"
```
Comment (emdupre) [inline maint_tools/check_gha_workflow.py:48]: Is this something that we have to set for you individually, or could it be set under the org ?

File: maint_tools/check_gha_workflow.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    else:
+        return "n/a"
+
+
+def set_python_version(x: str) -> str:
+    """Detect which python version the job was run on."""
+    return next(
+        (
+            version
+            for version in ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
Comment (emdupre) [inline maint_tools/check_gha_workflow.py:160]: Is there a way we could keep this in-sync with the Python versions we support, without having to remember to update it here ?

File: maint_tools/check_gha_workflow.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    if response.status_code != 200:
+        raise RuntimeError(
+            f"\n Got code: {response.status_code}.\n"
+            f" With response: {response.json()}"
+        )
+
+    return response.json()
+
+
+def udpate_jobs_data(
```
Comment (emdupre) [inline maint_tools/check_gha_workflow.py:244]: Will also need to update the associated call(s)

```suggestion
def update_jobs_data(
```

File: maint_tools/check_gha_workflow.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
+"""Collects and plot run time of jobs in a GHA workflow.
+
+Requires:
+- requests
+- rich
+- pandas
+- plotly
+- kaleido
```
Comment (emdupre) [inline maint_tools/check_gha_workflow.py:8]: Where is kaleido used in the script ?

Comment (Remi-Gau) [reply maint_tools/check_gha_workflow.py:8]: We need it to save the plotly images to disk.

File: maint_tools/check_gha_workflow.py (hunk 2 above)
Comment (Remi-Gau) [reply maint_tools/check_gha_workflow.py:48]: The username and the token file are there if you want or need to run the script locally.
In CI we use the org token directly.

I will improve the comment to better explain this.

File: maint_tools/check_gha_workflow.py (hunk 3 above)
Comment (Remi-Gau) [reply maint_tools/check_gha_workflow.py:160]: Cannot think of an easy way... Yet.

File: maint_tools/check_gha_workflow.py (hunk 5 above)
Comment (emdupre) [reply maint_tools/check_gha_workflow.py:8]: Got it, I see now that newer plotly versions (4.9+) will automatically use it when installed, but could we make that explicit ? So, set `engine="kaleido"`. I can add a suggestion below.

File: maint_tools/check_gha_workflow.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            "OS": "OS",
+            "python": "python version",
+        },
+        title="Duration of nilearn test runs",
+    )
+
+    fig.update_xaxes(dtick="M1", tickformat="%b\n%Y")
+    fig.update_layout(autosize=True, width=1000, height=700)
+
+    fig.write_image(OUTPUT_FILE.with_suffix(".png"))
```
Comment (emdupre) [inline maint_tools/check_gha_workflow.py:139]:
```suggestion
    fig.write_image(OUTPUT_FILE.with_suffix(".png"), engine="kaleido")
```

Comment (Remi-Gau) [reply maint_tools/check_gha_workflow.py:139]: thanks

File: maint_tools/check_gha_workflow.py (hunk 4 above)
Comment (Remi-Gau) [reply maint_tools/check_gha_workflow.py:244]: good catch
will fix the calls

File: maint_tools/check_gha_workflow.py (hunk 3 above)
Comment (Remi-Gau) [reply maint_tools/check_gha_workflow.py]: Could not find an easy way to do this, so for now I at least extract this list into a constant at the top of the script that should make it easier to find and modify.

File: maint_tools/check_gha_workflow.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+- pass the token directly to the script as an argument.
+  This is the way to do it when using this script in continuous integration.
+
+USAGE
+-----
+
+.. code-block:: bash
+
+    python maint_tools/check_gha_workflow.py $GITHUB_TOKEN
```
[the remark this answers is not in the document]
Comment (emdupre) [reply maint_tools/check_gha_workflow.py:38]: Yes, thanks !

**Pull request merged**